src/restic/backend/gs/gs.go

```diff
@@ -0,0 +1,461 @@
+func (fi fileInfo) Sys() interface{}   { return nil }        // underlying data source (can return nil)
+
+// ReadDir returns the entries for a directory.
+func (be *Backend) ReadDir(dir string) (list []os.FileInfo, err error) {
+	debug.Log("ReadDir(%v)", dir)
+
+	// make sure dir ends with a slash
+	if dir[len(dir)-1] != '/' {
+		dir += "/"
+	}
+
+	obj, err := be.service.Objects.List(be.bucketName).Prefix(dir).Delimiter("/").Do()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, item := range obj.Prefixes {
+		entry := fileInfo{
+			name:  strings.TrimPrefix(item, dir),
+			isDir: true,
+			mode:  os.ModeDir | 0755,
+		}
+		list = append(list, entry)
+	}
+	for _, item := range obj.Items {
+		entry := fileInfo{
+			name:  strings.TrimPrefix(item.Name, dir),
+			isDir: false,
+			mode:  0644,
+			size:  int64(item.Size),
+			//modTime: item.Updated,
+		}
+		if entry.name != "" {
+			list = append(list, entry)
+		}
+	}
+
+	return list, nil
+}
+
+// Location returns this backend's location (the bucket name).
+func (be *Backend) Location() string {
+	return be.Join(be.bucketName, be.prefix)
+}
+
+// Path returns the path in the bucket that is used for this backend.
+func (be *Backend) Path() string {
+	return be.prefix
+}
+
+// getRemainingSize returns number of bytes remaining. If it is not possible to
+// determine the size, panic() is called.
+func getRemainingSize(rd io.Reader) (size int64, err error) {
+	type Sizer interface {
+		Size() int64
+	}
+
+	type Lenner interface {
+		Len() int
+	}
+
+	if r, ok := rd.(Lenner); ok {
+		size = int64(r.Len())
+	} else if r, ok := rd.(Sizer); ok {
+		size = r.Size()
+	} else if f, ok := rd.(*os.File); ok {
+		fi, err := f.Stat()
+		if err != nil {
+			return 0, err
+		}
+
+		pos, err := f.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return 0, err
+		}
+
+		size = fi.Size() - pos
+	} else {
+		panic(fmt.Sprintf("Save() got passed a reader without a method to determine the data size, type is %T", rd))
+	}
+	return size, nil
+}
+
+// preventCloser wraps an io.Reader to run a function instead of the original Close() function.
+type preventCloser struct {
+	io.Reader
+	f func()
+}
+
+func (wr preventCloser) Close() error {
+	wr.f()
+	return nil
+}
+
+// Save stores data in the backend at the handle.
+func (be *Backend) Save(ctx context.Context, h restic.Handle, rd io.Reader) (err error) {
+	if err := h.Valid(); err != nil {
+		return err
+	}
+
+	objName := be.Filename(h)
+
+	size, err := getRemainingSize(rd)
+	if err != nil {
+		return err
+	}
+
+	debug.Log("Save %v at %v", h, objName)
+
+	// Check key does not already exist
+	if _, err := be.service.Objects.Get(be.bucketName, objName).Do(); err == nil {
+		debug.Log("%v already exists", h)
+		return errors.New("key already exists")
+	}
+
+	be.sem.GetToken()
+
+	// wrap the reader so that net/http client cannot close the reader, return
+	// the token instead.
+	rd = preventCloser{
```

Comment:
Is this needed for the gcs client? Does it really close the reader? You can run the integration tests, which checks that. Otherwise drop the preventCloser{}.